### PR TITLE
Interactive building for debugging

### PR DESCRIPTION
**Pull request by casparvl**
This PR moves the 'adding software' pages to a new top level header, as I was told by an external user he hadn't found it initially.

Also, added a page on how to interactively replicated the build process.

### CODE REVIEW

File: mkdocs.yml
Comment (trz42) [inline mkdocs.yml:43]: "Contributing" is maybe not the best wording? How about "Adding ..."? Then the first item should be renamed.

It could be something like
- Adding software to EESSI
  - Overview
  - Contribution policy (requires #108)
  - Opening a pull request (anyone)
  - Building software (maintainers)
  - Debugging build issues (anyone)
  - Deploying built packages (maintainers)

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [inline docs/contributing_sw/debugging_failed_builds.md:121]: Take out `$HOME`

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [inline docs/contributing_sw/debugging_failed_builds.md:124]: use `mktemp` instead to keep it unique, to avoid clashes between multiple users on the node

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [inline docs/contributing_sw/debugging_failed_builds.md:126]: Use WORKDIR so that you land in the same tempdir

File: docs/contributing_sw/debugging_failed_builds.md
Comment (trz42) [inline docs/contributing_sw/debugging_failed_builds.md:95]: Why this? Ah, because `startprefix` clears those environment variables?

Then, `run_in_compat_layer_env.sh` is the way to go, because it preserves these in some way.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (trz42) [inline docs/contributing_sw/debugging_failed_builds.md:98]: Hmm, do we need to state this here?

File: docs/contributing_sw/debugging_failed_builds.md
Comment (trz42) [inline docs/contributing_sw/debugging_failed_builds.md:110]: Possibly this could be removed if you run with `run_in_compat_layer_env.sh`

File: docs/contributing_sw/debugging_failed_builds.md
Comment (trz42) [inline docs/contributing_sw/debugging_failed_builds.md:119]: Unless one wants to build into `/cvmfs/...`

File: docs/contributing_sw/debugging_failed_builds.md
Comment (trz42) [inline docs/contributing_sw/debugging_failed_builds.md:121]: Should we prefer local disks for the working directory, so avoid parallel file systems? Unsure about that one. I'd say recommend what you do when using EasyBuild locally.

Comment (trz42) [reply docs/contributing_sw/debugging_failed_builds.md:121]: Ah, one thing, if you use something under `/tmp` it may be saved into the tarball created by the `eessi_container.sh` script. If so that might be beneficial.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (trz42) [inline docs/contributing_sw/debugging_failed_builds.md:178]:
```suggestion
After some time, this build fails while trying to build `Plumed`, and we can access the build log to look for clues on why it failed.
```

File: docs/contributing_sw/debugging_failed_builds.md
Comment (trz42) [inline docs/contributing_sw/debugging_failed_builds.md:181]: Maybe?
```suggestion
In our [example PR](https://github.com/EESSI/software-layer/pull/360), the individual package that was added to `eessi-2023.06-eb-4.8.1-2021b.yml` was `LAMMPS-23Jun2022-foss-2021b-kokkos.eb`. We'll also have to (re)use any options that are listed in the EasyStack file for `LAMMPS-23Jun2022-foss-2021b-kokkos.eb`, in this case the option `--from-pr 19000`. Thus, to build, we run:
```

File: docs/contributing_sw/debugging_failed_builds.md
Comment (trz42) [inline docs/contributing_sw/debugging_failed_builds.md:185]:
```suggestion
After some time, this build fails while trying to build `Plumed`, and we can access the build log to look for clues on why it failed.
```

File: mkdocs.yml
Comment (casparvl) [reply mkdocs.yml:43]: One thing I'm wondering: from 'Adding software to EESSI', is it immediately clear that this is about _anyone_ being able to do that? From that header, it could also be a page that just describes how _we_ add software to EESSI.

The nice thing about the word `contributing` is that it is immediately clear that any contributor can do that.

I do like your other headers, and the suggestion to include the contribution policy under this new header.

Comment (casparvl) [reply mkdocs.yml:43]: Never mind, I'm bikeshedding. Adding is fine, I'll do that :D

I've created this structure, with 'coming soon' pages for the things we don't have content for yet. At least the full structure is in place like this.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:121]: Done

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:126]: Done

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:5]: Rephrased.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:9]: Rephrased

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:49]: Changed

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:51]: Ok, but since we later on suggest to put our `WORKDIR` and `EASYBUILD_INSTALLPATH` in `/tmp`, I think it makes sense to keep this instruction. I've elaborated a bit in that it is specifically meant for retaining the `/tmp` data, which is the direction we later advise as workdir & installdir. Hopefully people can then make the connection themselves that if they use a different `WORKDIR` and `EASYBUILD_INSTALLDIR`, that they don't need to use `--save`.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:54]: I'll add the remark that the tarballs can be sizeable, and that they should pick an FS where they have enough quotum.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:68]: fixed.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md]: Hm, so I now used the `minimal_eessi_env`  script or whatever its called. But you're right, we can probably just extract the two lines that are needed from this script. That makes it more transparent what's happening, and also less dependency on this script to stay around.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:95]: We discussed this, it's not an option since `run_in_compat_layer_env.sh` has no option of dropping you in an interactive shell - it runs some commands and then returns to a shell outside of prefix.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md]: Removed it, as we now set it using `uname -m` (and linux, hardcoded).

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:98]: Well, it gives some idea of _why_ we do this, and also provides a way for users who are in doubt to easily check if they are now in a prefix environment. I think as a note, it's ok.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:110]: See above

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:119]: Agreed. I've had one issue that didn't show up when building somewhere else. But: I don't know what needs to change exactly to build in `/cvmfs/`. I guess we need to pass `--access rw` then? Does it then mount the writeable overlay? Or is there more that needs to be done?

As I understand, it is a bit slower to compile in the writeable overlay. What would be your idea: should we advise that by default, simply to stay as close to the bot as possible? Or not, because it is slower?

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:121]: Yeah, it is. I've tried that combined with `--save`. Indeed, I could easily resume, having some dependencies already build in `/tmp` the first time. Saved me time :)

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:119]: Also, if we install there, is that saved into the tarball as well by `--save` and then restored by `--restore`? I.e. if I need to build some deps first,, then save, then restore later, will the deps be already there?

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:178]: Done

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:181]: Done

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:185]: Done

File: docs/contributing_sw/debugging_failed_builds.md
Comment (trz42) [inline docs/contributing_sw/debugging_failed_builds.md:184]: Maybe one should caution here, that the bot is "never" running `eb` to install packages like that. So, the command could just work and not fail simply because it is just a little different.

We hit such false positives when debugging failing builds of `PROJ-9.0.0-GCCcore-11.3.0.eb`. The error message in the bot's EB build log was not telling much, and instead of processing the easystack file we just ran the command to install a single `.eb` file in our debugging session.

**Review by trz42 — CHANGES_REQUESTED** [review]
Two more suggestions.

I think we should wait merging this PR until #108 has been merged and a short overview (`overview.md`) is added.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:125]: I'll make it `eessi-debug.XXXXXXX`. It's still good to know it is related to EESSI, in case you find a stray dir like this in your `/tmp`.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:184]: Done, put a note with this caveat.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md]: Ok, `--access rw` and _not_ setting `EASYBUILD_INSTALLPATH` ourselves (i.e. relying on the one set by `configure_easybuild` works. This creates builds that actually get installed into the writeable overlay. That has the advantage of being even more similar to what the bot does. Hence, I was able to replicate also [this issue](https://github.com/EESSI/software-layer/pull/332#issuecomment-1775374260).

I think that is an improvement. I still add to the instructions the suggestion of setting an `EASYBUILD_INSTALLPATH`, but only _after_ having replicated the issue - as a way of identifying if the writeable overlay is somehow causing an issue.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (t1mk1k) [inline docs/contributing_sw/debugging_failed_builds.md:92]:
```suggestion
Now, reset the `${EESSI_CVMFS_REPO}` and `${EESSI_PILOT_VERSION}` in your prefix environment with the initial values (printed in the echo statements above)
```

File: docs/contributing_sw/debugging_failed_builds.md
Comment (t1mk1k) [inline docs/contributing_sw/debugging_failed_builds.md:112]:
```suggestion

!!! Note
```

Extra line so notes are not on the same line

File: docs/contributing_sw/debugging_failed_builds.md
Comment (t1mk1k) [inline docs/contributing_sw/debugging_failed_builds.md:130]: Might be good to add this step after going describing the initial debug run. To us copy-paste monkeys, this one looks like one of the commands you should use, if you skim (or fail to read) the text above.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (t1mk1k) [inline docs/contributing_sw/debugging_failed_builds.md:133]:
```suggestion
(_after_ sourcing the `configure_easybuild` script, so that you overwrite whatever that script sets). This can help you identify if an issue is related to building in a writeable overlay. For example, the writeable overlay is known to be a bit slow sometimes, and we have seen tests failing because they exceeded some timeout.
```

File: docs/contributing_sw/debugging_failed_builds.md
Comment (t1mk1k) [inline docs/contributing_sw/debugging_failed_builds.md:136]:
```suggestion
    If you started the container using --resume, you may want WORKDIR to point to the workdir you created previously (instead of creating a new, temporary directory with `mktemp`).
```

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:92]: Changed

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:112]: Changed

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:130]: Agreed, I made a new section on known causes of issues (after the 'build' instructions), and moved it there.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:133]: Rephrased this whole section, so this typo is gone now.

File: docs/contributing_sw/debugging_failed_builds.md
Comment (casparvl) [reply docs/contributing_sw/debugging_failed_builds.md:136]: fixed

**Review by trz42 — APPROVED** [review]

**Comment by trz42** [comment]
Looks good to me. Thanks @casparvl !

**Pull request merged**
